### ISSUE TITLE
Suitable Constructor Type for SCP Could Not Be Located

### ISSUE THREAD

**Issue by mschmidt73** [fo-dicom/fo-dicom#1611]
**Describe the bug**
When a DicomServer<T> instance is started a needs to create it's SCP the following exception is thrown:

System.InvalidOperationException: A suitable constructor for type 'DicomProxyServer.Services.UpstreamDicomService' could not be located. Ensure the type is concrete and all parameters of a public constructor are either registered as services or passed as arguments. Also ensure no extraneous arguments are provided.
   at Microsoft.Extensions.DependencyInjection.ActivatorUtilities.FindApplicableConstructor(Type instanceType, Type[] argumentTypes, ConstructorInfo& matchingConstructor, Nullable`1[]& matchingParameterMap)
   at Microsoft.Extensions.DependencyInjection.ActivatorUtilities.CreateFactory(Type instanceType, Type[] argumentTypes)
   at FellowOakDicom.Network.DicomServer`1.CreateScp(INetworkStream stream)

**To Reproduce**
This seems to have started with v 5.1.x. I originally discovered this in a brand new .NET 7 WebApi (minimal) project I created with v 5.1.1 installed from the start. We had a project (based off of my POC) that was incorrectly started as a worker template, rather than an API. I created the new project and copied code into it that was working in v 5.0.3. After converting over from the old FO ILog to Micorosoft.Extensions.Logging.ILogger, I started seeing this exception.

I tried debugging to test whether all dependencies could be created from an IServiceProvider, and the could be (through I switched a couple of DI registrations from Scoped to Transient, just in case). I then tried using the old UserState object to pass in dependencies other than INetworkStream, Encoding, ILogger<T> and DicomServiceDependencies - same thing.

Lastly, I went back to my functioning POC server project (which has been working on 5.0.3 for ~6 months now) and upgraded it from 5.0.3 -> 5.1.1. I updated the code by removing the old LogManager I had configured (to point to ILogger) and switched over from FO logging to MSFT logging - same thing.

**Expected behavior**
I would expected 5.0.3 SCP code to function pretty much exactly the same in 5.1.1 once converted over to using Microsoft.Extensions.Logging. I would also expect to be able to use standard DI constructor injections of other dependencies but, if necessary, we can continue using UserState (as I have been in the POC) as a fallback.

**Screenshots or test DICOM files**
If applicable, add screenshots or test DICOM files to help explain your problem.

**Environment**
Fellow Oak DICOM version: 5.1.1
OS: MacOS (but irrelevant)
Platform: .NET 7

**Comment by amoerie** [comment]
First thought that comes to mind: is it possible you are using the static DicomServerFactory.Create method instead of injecting an `IDicomServerFactory` in the place where you create the server?

I just tested with the latest release and this sample works:

**Sample.csproj**

```xml
<Project Sdk="Microsoft.NET.Sdk.Worker">

    <PropertyGroup>
        <OutputType>Exe</OutputType>
        <TargetFramework>net7.0</TargetFramework>
        <ImplicitUsings>enable</ImplicitUsings>
        <Nullable>enable</Nullable>
    </PropertyGroup>

    <ItemGroup>
      <PackageReference Include="fo-dicom" Version="5.1.1" />
      <PackageReference Include="Microsoft.Extensions.Hosting" Version="7.0.1" />
    </ItemGroup>
</Project>
```

**Program.cs**

```csharp
using System.Text;
using FellowOakDicom;
using FellowOakDicom.Network;

var host = Host.CreateDefaultBuilder(args)
    .ConfigureLogging(logging =>
    {
        logging.ClearProviders();
        logging.AddConsole();
        logging.SetMinimumLevel(LogLevel.Information);
    })
    .ConfigureServices(services =>
    {
        services.AddFellowOakDicom();
        services.AddHostedService<Worker>();
        // This will be injected into the DICOM service
        services.AddSingleton<CustomDependency>();
    })
    .Build();

DicomSetupBuilder.UseServiceProvider(host.Services);

host.Run();

public class CustomDependency
{
    
}

public class Worker : IHostedService
{
    private readonly ILogger<Worker> _logger;
    private readonly IDicomServerFactory _dicomServerFactory;
    private IDicomServer? _server;

    public Worker(ILogger<Worker> logger, IDicomServerFactory dicomServerFactory)
    {
        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
        _dicomServerFactory = dicomServerFactory ?? throw new ArgumentNullException(nameof(dicomServerFactory));
    }

    public async Task StartAsync(CancellationToken cancellationToken)
    {
        _logger.LogInformation("Starting DICOM server");
        _server = _dicomServerFactory.Create<EchoService>(104);
        _logger.LogInformation("DICOM server is running");
    }

    public Task StopAsync(CancellationToken cancellationToken)
    {
        if (_server != null)
        {
            _server.Stop();
            _server.Dispose();
            _server = null;
        }
        return Task.CompletedTask;
    }
}

public class EchoService : DicomService, IDicomServiceProvider, IDicomCEchoProvider
{
    private readonly ILogger<EchoService> _logger;
    private readonly CustomDependency _customDependency;

    public EchoService(INetworkStream stream,
        Encoding fallbackEncoding, DicomServiceDependencies dependencies,
        ILogger<EchoService> logger,
        CustomDependency customDependency) : base(stream, fallbackEncoding, logger, dependencies)
    {
        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
        _customDependency = customDependency ?? throw new ArgumentNullException(nameof(customDependency));
    }

    public void OnReceiveAbort(DicomAbortSource source, DicomAbortReason reason) => _logger.LogInformation("Received abort");
    public void OnConnectionClosed(Exception exception) => _logger.LogInformation("Connection closed"); 
    public Task OnReceiveAssociationRequestAsync(DicomAssociation association) => SendAssociationAcceptAsync(association);
    public Task OnReceiveAssociationReleaseRequestAsync()
    {
        _logger.LogInformation("Received association release");
        return Task.CompletedTask;
    }

    public Task<DicomCEchoResponse> OnCEchoRequestAsync(DicomCEchoRequest request) => Task.FromResult(new DicomCEchoResponse(request, DicomStatus.Success));
}
```

Output

```log
info: Worker[0]
      Starting DICOM server
info: Worker[0]
      DICOM server is running
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
```

**Comment by mschmidt73** [comment]
I'm trying this with the webapi (minimal) template. It would seem strange, but could it be an issue with the different host?

I've omitted a bunch of stuff for brevity. I also only tried this with one simple singleton dependencies.

```
var builder = WebApplication.CreateBuilder(args);

var appSettings = builder.RegisterAndValidateAppSettings();

builder.Services.AddFellowOakDicom();

builder.Services.AddHostedService<DicomServerHostedService>();

var app = builder.Build();

DicomSetupBuilder.UseServiceProvider(app.Services);

app.Run();
```

```
public class DicomServerHostedService : IHostedService
{
    private readonly AppSettings _appSettings;
    private readonly IDicomServerFactory _factory;
    private readonly ILogger<DicomServerHostedService> _logger;

    private IDicomServer? _dicomServer;

    public DicomServerHostedService(IDicomServerFactory factory, AppSettings appSettings, ILogger<DicomServerHostedService> logger)
    {
        _factory = factory;
        _appSettings = appSettings;
        _logger = logger;
    }

    public virtual Task StartAsync(CancellationToken cancellationToken)
    {
        if (!cancellationToken.IsCancellationRequested)
        {
            _logger.LogInformation("Starting DICOM Server ...");
            _logger.LogInformation("Server: {0}. Port: {1}", _appSettings.Server.Host, _appSettings.Server.Port);

            _dicomServer = _factory.Create<UpstreamDicomService>(_appSettings.Server.Host, _appSettings.Server.Port);
        }
        
        return Task.CompletedTask;
    }

    public virtual Task StopAsync(CancellationToken cancellationToken)
    {
        _dicomServer?.Stop();
        return Task.CompletedTask;
    }
}
```

```
public class UpstreamDicomService : DicomService, IDicomCEchoProvider, IDicomCFindProvider, IDicomCGetProvider, IDicomCMoveProvider, IDicomCStoreProvider, IDicomServiceProvider
{
    private readonly IDicomAdapterClient _adapterClient;
    private readonly LiteQueue<CreateInstance> _createInstanceQueue;
    private readonly IDicomClient _client;
    private readonly string _facilityCode;

    public UpstreamDicomService(
        INetworkStream stream,
        Encoding fallbackEncoding,
        DicomServiceDependencies dependencies,
        ILogger<UpstreamDicomService> logger,
        AppSettings appSettings) : base(stream, fallbackEncoding, logger, dependencies)
     {
        _facilityCode = appSettings.FacilityCode;
     }
}
```

**Comment by mschmidt73** [comment]
Also, did you try running an actual C-ECHO command against the server? I get the same output but the server won't try to create the SCP class until it receives it's first command. (At least that's my understanding)

**Comment by amoerie** [comment]
Are you sure all dependencies are registered and resolvable? 
It would help if you could share a sample that fully compiles so I can debug through it. 

So far, the only likely explanation I can come up with is that one of your services is not registered, or that you did not call .AddLogging so that ILogger is resolvable.

**Comment by amoerie** [comment]
> Also, did you try running an actual C-ECHO command against the server? I get the same output but the server won't try to create the SCP class until it receives it's first command. (At least that's my understanding)

Ah that's a good point. Hang on.

**Comment by mschmidt73** [comment]
Logging (we have no issues resolving ILogger<T> elsewhere in the project) ...

```
builder.Logging.ClearProviders();
builder.Logging.AddConsole();
builder.Logging.AddOpenTelemetry(configure =>
{
    var myExporter = builder.Services.BuildServiceProvider().GetRequiredService<LogExporter>();
    configure.AddProcessor(new SimpleLogRecordExportProcessor(myExporter));
});
```

**Comment by amoerie** [comment]
I can reproduce the issue after actually sending a request. D'oh. I'll keep you posted. :-)

**Comment by amoerie** [comment]
I found the root cause.

ActivatorUtilities is very finicky and requires an exact type match. In a nutshell, Fellow Oak DICOM specifies that every DicomService must have these exact 3 parameters in the constructor signature: `INetworkStream stream, Encoding fallbackEncoding, ILogger logger` (the names don't matter)

What we both did was change this `ILogger` to be of type `ILogger<T>`, which breaks the contract. (even though `ILogger<T>` extends `ILogger`)

Edit: here is a fully working example:

```csharp
using System.Text;
using FellowOakDicom;
using FellowOakDicom.Network;
using FellowOakDicom.Network.Client;

var host = Host.CreateDefaultBuilder(args)
    .ConfigureLogging(logging =>
    {
        logging.ClearProviders();
        logging.AddConsole();
        logging.SetMinimumLevel(LogLevel.Information);
    })
    .ConfigureServices(services =>
    {
        services.AddFellowOakDicom();
        services.AddHostedService<Worker>();
        // This will be injected into the DICOM service
        services.AddSingleton<CustomDependency>();
    })
    .Build();

DicomSetupBuilder.UseServiceProvider(host.Services);

host.Run();

public class CustomDependency
{
    
}

public class Worker : IHostedService
{
    private readonly ILogger<Worker> _logger;
    private readonly IDicomServerFactory _dicomServerFactory;
    private readonly IDicomClientFactory _dicomClientFactory;
    private IDicomServer? _server;

    public Worker(ILogger<Worker> logger, IDicomServerFactory dicomServerFactory, IDicomClientFactory dicomClientFactory)
    {
        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
        _dicomServerFactory = dicomServerFactory ?? throw new ArgumentNullException(nameof(dicomServerFactory));
        _dicomClientFactory = dicomClientFactory ?? throw new ArgumentNullException(nameof(dicomClientFactory));
    }

    public async Task StartAsync(CancellationToken cancellationToken)
    {
        _logger.LogInformation("Starting DICOM server");
        _server = _dicomServerFactory.Create<EchoService>(104);
        _logger.LogInformation("DICOM server is running");

        var client = _dicomClientFactory.Create("127.0.0.1", 104, false, "AnySCU", "AnySCP");

        _logger.LogInformation("Sending C-ECHO request");
        DicomCEchoResponse? response = null;
        await client.AddRequestAsync(new DicomCEchoRequest { OnResponseReceived = (_, r) => response = r});
        await client.SendAsync(cancellationToken);
        if (response != null)
        {
            _logger.LogInformation("C-ECHO response received");
        }
        else
        {
            _logger.LogError("No C-ECHO response received");
        }
    }

    public Task StopAsync(CancellationToken cancellationToken)
    {
        if (_server != null)
        {
            _server.Stop();
            _server.Dispose();
            _server = null;
        }
        return Task.CompletedTask;
    }
}

public class EchoService : DicomService, IDicomServiceProvider, IDicomCEchoProvider
{
    private readonly ILogger _logger;
    private readonly CustomDependency _customDependency;

    public EchoService(INetworkStream stream,
        Encoding fallbackEncoding, 
        ILogger logger,
        DicomServiceDependencies dependencies,
        CustomDependency customDependency) : base(stream, fallbackEncoding, logger, dependencies)
    {
        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
        _customDependency = customDependency ?? throw new ArgumentNullException(nameof(customDependency));
    }

    public void OnReceiveAbort(DicomAbortSource source, DicomAbortReason reason) => _logger.LogInformation("Received abort");
    public void OnConnectionClosed(Exception exception) => _logger.LogInformation("Connection closed");

    public Task OnReceiveAssociationRequestAsync(DicomAssociation association)
    {
        foreach (DicomPresentationContext presentationContext in association.PresentationContexts)
            presentationContext.SetResult(DicomPresentationContextResult.Accept);
        return SendAssociationAcceptAsync(association);
    }

    public Task OnReceiveAssociationReleaseRequestAsync()
    {
        _logger.LogInformation("Received association release");
        return Task.CompletedTask;
    }

    public Task<DicomCEchoResponse> OnCEchoRequestAsync(DicomCEchoRequest request) => Task.FromResult(new DicomCEchoResponse(request, DicomStatus.Success));
}
```

P.S. If you absolutely want a `ILogger<T>`, you can make that happen by just adding it as an extra constructor parameter:

```csharp
    private readonly ILogger<EchoService> _logger;
    private readonly CustomDependency _customDependency;

    public EchoService(INetworkStream stream,
        Encoding fallbackEncoding, 
        ILogger baseLogger,
        DicomServiceDependencies dependencies,
        CustomDependency customDependency,
        ILogger<EchoService> logger) : base(stream, fallbackEncoding, baseLogger, dependencies)
    {
        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
        _customDependency = customDependency ?? throw new ArgumentNullException(nameof(customDependency));
    }
```

**Comment by amoerie** [comment]
We might want to document somewhere that `ILogger` cannot be changed to `ILogger<T>`. I'm not sure if we can magically make this scenario work... I suppose we could use reflection to inspect the constructor beforehand, so we could support subtypes of `ILogger`.

**Comment by mschmidt73** [comment]
Sorry it took a minute to respond to this - just got back from a trip. I was able to get things working with version 5.1.1 using the factory and changing from `ILogger<T>` to `ILogger`. I'm pretty sure our team will prefer to inject `ILogger<T>` so I would be willing to take a look at adding this functionality.

Note from @amoerie: I edited your comment to fix the formatting of `ILogger<T>`, because it was a bit confusing
